Add support for multiple clusters

kaiyou opened the pull request:
This PR mostly implements updates for #38 and #40. 

A first intuitive implementation was trying to add a `cluster.name` field, used to fill the state path dynamically. This ended up in a mess of code and no clear use for the cluster name except for the state path and banner.

Instead, this PR simply expose both as config options directly, named `--banner` and `--state-path`. It also adds an e2e test running two clusters that share a node.

It is still missing documentation, hence the WIP flag.

costela commented on main.go:
```diff
@@ -46,6 +46,7 @@ func main() {
 
 	// Prepare the /etc/hosts writer
 	hostsFile := &etchosts.EtcHosts{
+		Banner: "# ! managed automatically for wesher " + config.Banner,
```
```suggestion
		Banner: "# ! managed automatically by wesher " + config.Banner,
```

costela commented on config.go:
```diff
@@ -14,6 +14,8 @@ type config struct {
 	ClusterKey    []byte   `id:"cluster-key" desc:"shared key for cluster membership; must be 32 bytes base64 encoded; will be generated if not provided"`
 	Join          []string `desc:"comma separated list of hostnames or IP addresses to existing cluster members; if not provided, will attempt resuming any known state or otherwise wait for further members."`
 	Init          bool     `desc:"whether to explicitly (re)initialize the cluster; any known state from previous runs will be forgotten"`
+	Banner        string   `id:"banner" desc:"a string appended to /etc/hosts banner to discriminate between multiple running clusters"`
+	StatePath     string   `id:"state-path" desc:"path where the cluster state is stored; optional, depends on cluster name if not set"`
```
I'm not sure if we need to expose these as settings explicitly. 
How about adding a cluster name setting and explicitly mentioning it's potential use?
```suggestion
	ClusterName        string   `id:"cluster-name" desc:"identifier for the wesher cluster; can be used to peer with multiple clusters" default:"default"`
```

kaiyou replied:
That was my first intention and implementation. However, one still needs to specify different ports, overlay networks, etc. so configuration is not not as trivial as setting the cluster name, plus one needs to carry the cluster name all the way to `state.save` and `loadState`.

I will reapply my commits on the matter so you can compare as well and we'll follow your preference.

costela replied:
Now that I think about it a bit more, I wonder if we can't get rid of `ClusterName` as well.
As you mentioned, `ClusterPort` already need to be set for multiple instances to coexist. I'm not sure about `OverlayNet` (it's theoretically possible to have two clusters on the same net, as long as the hostnames are disjoint; maybe not the best idea, but possible).

So maybe we can just use these existing settings to determine banner and state file?

costela replied:
The question for me seems to be: "what do I need to do, to run multiple wesher instances?"
And the answer should be "use different cluster ports (and probably overlay nets)"

We don't need to give the user the footguns for setting banner and state file path. These are implementation details that don't help him and would only mean more settings to change.

Does that make sense to you?

kaiyou replied:
Interface should also be different, since private key is per wireguard interface. Then so should be the wireguard port.

Maybe it makes more sense to use the interface name? It would also make things more humanly readable, you'll find your cluster state in `/var/lib/wesher/mywg.json` and `/etc/hosts` will clearly state which interface is responsible for a host.

costela replied:
Yes, that makes more sense.
Ideally we'd als document the fact that multiple instances need separate:
- wg interface
- wg port
- cluster port
- (probably) overlay net

kaiyou replied:
OK. Then do we ensure at least some backward compatibility, and save to the current path if no interface name is specified?

We will break backward compatibility anyway for users that already specify an interface name.

costela replied:
Not if we use the old path as a fallback on load.
- try to load the new path
- if not found try the old one
- save to the new one

kaiyou replied:
That should be okay indeed. Will do, thanks!

kaiyou commented:
Thank you for the review!

Just pushed a version exposing `ClusterName` instead, with related documentation.

costela commented on cluster/state.go:
```diff
@@ -16,10 +17,10 @@ type state struct {
 
-// TODO: this should be replaced by a configurable path later
-var statePath = "/var/lib/wesher/state.json"
+var defaultStatePath = "/var/lib/wesher/%s.json"
```
Changing the state file name has the unfortunate consequence of breaking existing clusters on upgrade. 

I think we have to add some special case handling for the "default" cluster and also try to load the old state file (`state.json`) if the new one (`default.json`) doesn't exist. Ugly, but rather this than breaking people's clusters.
And just for completeness' sake: calling the default cluster "state" feels even uglier :wink:

kaiyou replied:
Waiting for the other comment thread to be resolved since it will impact this as well.

kaiyou commented:
I removed the `cluster-name` settings in favor of using the interface name directly. Tests seem to be running fine and I feel okay about the new behavior. I also updated the docs about running multiple clusters.

costela commented on README.md:
```diff
@@ -146,6 +146,16 @@ All options can be passed either as command-line flags or environment variables:
 
+## Running multiple clusters
+
+Wesher supports having a node be a member of multiple clusters. The following settings must be explicitely set to non-conflicting values:
+- `--interface`, so that wireguard does not mix up traffic, the interface name is also used for naming the state file
+- `--cluster-port` or `--bind-addr` or `--bind-iface`, so that wesher can properly bind to its port
+- `--wireguard-port`, so that wireguard can properly bind to its port
+
+One should also use different overlay networks. Even though using the same network on different devices will work in some cases, it is not the intended use.
```
```suggestion
```suggestion
To make a node be a member of multiple clusters, simply start multiple wesher instances. The following settings must be explicitly set to different values for each instance:
- `--interface`
- either `--cluster-port`, or `--bind-addr` or `--bind-iface`
- `--wireguard-port`
- `--overlay-net` (optional, but recommended to avoid node address conflicts)
```
```

costela commented on tests/e2e.sh:
```diff
@@ -100,6 +100,35 @@ test_cluster_simultaneous_start() {
+    cluster1='--cluster-port 7946 --wireguard-port 51820 --interface wgoverlay --overlay-net 10.10.0.0/16'
+    cluster2='--cluster-port 7947 --wireguard-port 51821 --interface wgoverlay2 --overlay-net 10.11.0.0/16'
+    setup_wireguard='wireguard-go wgoverlay2 2>/dev/null >/dev/null'
+    join_cluster2="nohup /app/wesher --cluster-key 'ILICZ3yBMCGAWNIq5Pn0bewBVimW3Q2yRVJ/Be+b1Uc=' --join test2-orig $cluster2 2>/dev/null &"
```
Don't wanna ask too much of someone contributing their free-time, but can we maybe extend `tests/entrypoint.sh` to get the interface name out of `$@`? :sweat_smile: 
This way we can reuse it and keep the tests a bit cleaner.

Also, I suspect the `nohup` might be what's causing the tests to time-out on travis.

kaiyou replied:
I removed the nohup and added an argument parsing routine in entrypoint.sh.

costela commented on tests/e2e.sh:
```diff
@@ -100,6 +100,35 @@ test_cluster_simultaneous_start() {
+
+    docker stop test3-orig
+    docker start test3-orig
+    docker exec test3-orig bash -c "$setup_wireguard; $join_cluster2"
```
We probably want to restart it _without_  the explicit `--join` flag, right? To test the loading of the state file?

Now that I mention it, `test_node_restart` is also not testing what I wanted it to test. I should probably fix that :unamused:

kaiyou replied:
Done in the latest commit :)

kaiyou commented:
After my latest additions, tests will require that you rebuild the test image probably. It should be backward compatible with tests on the current master branch.

costela commented:
Thanks again!
I think we have to split the `entrypoint.sh` changes into a separate PR, otherwise the image used during testing won't have it.

Either that, or we build the test image locally on travis, which is a bit more work, but might be the better solution.

kaiyou commented:
I was not aware the image was built using travis. I will add them to a separate PR.

kaiyou commented:
I do not undetstand why builds time out. I guess the faulty command is the explicit exec on the entrypoint, then the additional wesher process.

I am guessing wesher is not down or stuck, but simply simply retrying with exponential backoff.

Now, what is different between my dev environment and Travis that makes it fail to join the cluster?

costela commented:
Yeah, I was wondering the same thing. Maybe we should add a bit more print-debugging, just to see exactly where the build is hanging.
I'm sorry that I can't devote more time for helping with this myself right now, but I'm still a bit swamped :unamused:

kaiyou commented:
It is my PR and my responsibility making things right, no worries :) Since you are okay with it, I'll add a couple debugging things to the PR for Travis to run, I'll force push later after removing them from history if all things are right.

kaiyou commented:
Okay, it is not an issue with the additional wesher joining its cluster. My new asumption it is somehow related to the lack of tty on Travis and the way docker exec behaves in that kind of environment. The wesher command does not return, so maybe docker exec does not return either, despite `&` and even `nohup` in my latest test.

kaiyou commented:
Now that tests are running fine, is there anything I can improve before we merge this?

costela commented:
thanks again for all the great work and patience with the slow reviews! :+1:
